File: core/MinimumIndexSumOfTwoLists/MinimumIndexSumOfTwoLists.java
package com.dbc;

import java.util.*;

public class MinimumIndexSumOfTwoLists {
    public String[] findRestaurant(String[] list1, String[] list2) {
        Map<String, Integer> remain = new HashMap<>();
        for (int i = 0; i < list1.length; i++) {
            if (remain.containsKey(list1[i])) continue;
            remain.put(list1[i], i);
        }

        List<String> res = new ArrayList<>();
        int min = list1.length + list2.length;
        for (int i = 0; i < list2.length; i++) {
            if (remain.containsKey(list2[i])) {
                if (i + remain.get(list2[i]) < min) {
                    res.clear();
                    res.add(list2[i]);
                    min = i + remain.get(list2[i]);
                } else if (i + remain.get(list2[i]) == min) {
                    res.add(list2[i]);
                }
            }
        }
        return res.toArray(new String[0]);
    }
}

File: core/MinimumIndexSumOfTwoLists/MinimumIndexSumOfTwoLists.py
from typing import List


def findRestaurant(list1: List[str], list2: List[str]) -> List[str]:
    remain = {}
    for index, item in enumerate(list1):
        if item in remain:
            continue
        remain[item] = index

    res, min_sum = [], len(list1) + len(list2)
    for index, item in enumerate(list2):
        if item in remain:
            if index + remain[item] < min_sum:
                min_sum = index + remain[item]
                res = [item]
            elif index + remain[item] == min_sum:
                res.append(item)

    return res


if __name__ == '__main__':
    print(findRestaurant(list1=["Shogun", "Tapioca Express", "Burger King", "KFC"],
                         list2=["KFC", "Shogun", "Burger King"]))
